### PR TITLE
Nlp merge

### PR DESCRIPTION
**Pull request by VivienZhou**
The files are meant to implement NLP service. NLP service is allowed to run with or without connecting with the command center. Please refer to readme for detailed usage information.

The service incorporates with Tim's structural changes now.

### CODE REVIEW

**Comment by jhauswald** [comment]
Hey can you [rebase](https://github.com/clarityinc/clarityinc-wiki/wiki/Adding-your-Feature#submit-a-pull-request-pr)? Look at @Lilisys PR #38 #40 for some inspiration.
I haven't looked through the code yet but its tough to follow given the commit history, the commits are helpful in understanding what's going on.

**Comment by VivienZhou** [comment]
Sure!
I have followed the instructions to rebase, but I will let you know as soon
as I rebase it.

On Monday, July 20, 2015, Johann Hauswald <notifications@github.com> wrote:

> Hey can you rebase
> <https://github.com/clarityinc/clarityinc-wiki/wiki/Adding-your-Feature#submit-a-pull-request-pr>?
> Look at @Lilisys <https://github.com/Lilisys> PR #38
> <https://github.com/clarityinc/clarityeco/pull/38> #40
> <https://github.com/clarityinc/clarityeco/pull/40> for some inspiration.
> I haven't looked through the code yet but its tough to follow given the
> commit history, the commits are helpful in understanding what's going on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clarityinc/clarityeco/pull/41#issuecomment-123122240>.
>


-- 
*Tiantong Zhou 周天童*
Computer Science, B.S.E. 2016
College of Engineering, University of Michigan
Electrical and Computer Engineering, B.S.E. 2016
UM-SJTU Joint Institute, Shanghai Jiao Tong University
tiantong@umich.edu | 1(734)680-4737

**Comment by VivienZhou** [comment]
Hey Johann, I have rebased the PR. Could you please check it? Thanks.

**Comment by VivienZhou** [comment]
Delete .the folder /templates

File: tonic-nlp/Makefile [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+CXX = g++
+CXX_FLAGS  = -O3 \
+                         -w \
+                         -fpermissive \
+                         -I$(COMMON)/include \
+                         -I$(CAFFE)/include \
+                         -I$(CUDA)/include \
+                         -Iz \
+                         -I/usr/local/include\
+                         -g
```
Comment (jhauswald) [inline tonic-nlp/Makefile]: no `-g`! and what is `-Iz`? also `make` prints `g++ -I-I/usr/local/include` which should be fixed.

File: tonic-nlp/Makefile [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+ifeq ($(CPU_ONLY), 1)
+    CXX_FLAGS  += -DCPU_ONLY
+endif
+
+LINK_FLAGS = $(LIBRARY_DIRS) \
+                         $(COMMON)/libtonic.a \
+                         -lopencv_core \
```
Comment (jhauswald) [inline tonic-nlp/Makefile]: tonic-nlp doesn't need any opencv so could you get rid of all these? Makefiles should only have what's necessary as they are already confusing, let's not add to that confusion.

File: tonic-nlp/Makefile [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+#File Names
+NLP_SERVICE  	= NLPService		
+NLP_CLIENT  	= NLPClient
+SOURCES = $(wildcard $(SRC)/*.cpp) $(wildcard $(GEN)/*.cpp)
+NLP_SERVICE_OBJ = $(SOURCES:.cpp=.o) NLPService.o		
+NLP_CLIENT_OBJ 	= $(SOURCES:.cpp=.o)  NLPTestClient.o
+
+# Main target
+#$(EXEC): $(OBJECTS) Makefile ;  $(CXX) $(OBJECTS) -o $(EXEC)  $(LIBS) $(LINK_FLAGS)
```
Comment (jhauswald) [inline tonic-nlp/Makefile]: delete

File: tonic-nlp/NLPService.cpp [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+
+      if (app.djinn){
+        SENNA_POS_forward_basic(pos, tokens->word_idx, tokens->caps_idx,
+                                   tokens->suff_idx, app);
+
+      	string out_string = call_djinn(app.pl.req_name, app.pl.num * app.pl.size);
+      	istringstream cin(out_string);
+      	string buff;
+       	for (int i = 0; i < app.pl.num * pos->output_state_size; i++){
+        	cin>>buff;
```
Comment (jhauswald) [inline tonic-nlp/NLPService.cpp]: What's this cin doing? Seems its writing the data returned from djinn into a buffer. I thought cin read stuff from the command line?

**Comment by jhauswald** [comment]
- Some of the remarks I already have are made in #38 so check those out (`cout` and `LOG(INFO)`) we should just have the same across all services.
- Can you comment your code a little more? I know whats going on here because we've talked about it but it would be good to have comments to describe the sequence, cool stuff that's going on. [here](https://github.com/Lilisys/clarityeco/blob/nlpMerge/tonic-nlp/NLPService.cpp#L480)
- what's the final decision with test-clients and where will they be? I think what you currently have is good.

I'll probably have more comments/feedback once I can test all of these. Like I said, I'm going to test an image and audio, so make sure it can handle it accordingly.

File: tonic-nlp/NLPService.cpp (hunk 4 above)
Comment (VivienZhou) [inline tonic-nlp/NLPService.cpp]: istringstream cin(out_string);
//out_string is a string consisting of a series of data separated by blanks. 
string buff;
for (...){
   cin>>buff
}
//Then in the for loop, buff will read the data stored in cin one by one

File: tonic-nlp/Makefile (hunk 1 above)
Comment (VivienZhou) [inline tonic-nlp/Makefile]: I have deleted -g and -Iz. However, make still prints g++ -I-I/usr/local/include. What is the problem if make prints  "g++ -I-I/usr/local/include"?

**Comment by VivienZhou** [comment]
1. Use LOG(INFO) to replace cout and printf.
2. I add some comments to explain my codes.
3. test-clients will be in the tonic-nlp folder, as concurrently constructed.
4. What do you mean by "test an image and audio?" NLP service only handler text input files. The image and audio must be converted into a text form before being processed by NLP service

**Comment by jhauswald** [comment]
1-3: sounds good
4: I'm referring to what I mentioned last week. What if by some odd chance your service, receives an image or speech, would it crash and burn or gracefully handle it and return empty data? Same goes for #40 or all the services actually.

I'm not saying this is going to happen but we want this code to be as robust as possible.

**Comment by VivienZhou** [comment]
For Img Service, it may crash when converting input data from string back to an image. However, NLP service stores input data into a string and then uses the string directly without converting the string back to a text. Can I eliminate images and audios according to file extensions? But in this way, the input files of NLP service have to be .txt files.

**Comment by jhauswald** [comment]
The tonic-nlp service shouldn't have access to any file extension. Are you reading text from disk in the service? Similar to the confusion we had with @Lilisys and sending images? Maybe this isn't such a big deal and the service would just return non-sense but address my question, that's important.

**Comment by jhauswald** [comment]
Tested, merging into dev

**Comment by jhauswald** [comment]
Oh btw @VivienZhou I had to modify your `NLPTestClient.cpp` to set the correct cc ip and ccport, it wasn't reading correctly from the cl, could you fix that and commit directly?

**Comment by VivienZhou** [comment]
Hey Johann, I have fixed this command line problem with ```NLPTestClient.cpp```. and push it to the pull request. However, how can I commit it directly?

**Pull request merged**